### ISSUE TITLE
64bit updates?

### ISSUE THREAD

**Issue by Lucarda** [ericlyon/pd-fftease#15]
Hi,

I'm trying to compile this for Windows 64bit and I _can_ but I get warnings in the Pd console like:

```
 "an operation on the array 'enrich-buffer' in the patch 'enrich~-help.pd'
failed since it uses garray_getfloatarray while running 64-bit"
```

and when **closing** help patches while DSP is on I get crashes. 

Is this nice lib intended to not support 64bit Pd's?

Would you accept PR's? 

I can try changing the build system to pd-lib-builder.

**Comment by ericlyon** [comment]
Greetings,

I haven't looked at Windows compilation in years, and my only Windows
laptop just died, so it's not something I can directly check into. I did
make a 64-bit compatible version a few years ago, though I did promote it,
since Pd at the time (and I think still) only does 32-bit computation in
its "64-bit" version. But have a look at this
<https://github.com/ericlyon/lyonpotpourri3.0> code base and see if you
have any more luck with it. If you compile successfully, I'll be happy to
host your Windows binaries on my website (with attribution).

I don't have any problem with PRs, but I also basically have no idea how
Github works, so I might need some advice about how to incorporate any code
suggestions.

Best wishes,
Eric

**Comment by porres** [comment]
> Is this nice lib intended to not support 64bit Pd's?

I don't think so as FFTease is available for Purr Data and that is compiled for 64 bit processors.

Let's just make a clear distinction as to not confuse @ericlyon - we are talking here about the binaries being compiled to run on Pd compled for 64 bit processors, we're not talking about double (64 bit) precision

**Comment by porres** [comment]
> I did make a 64-bit compatible version a few years ago, though I did promote it,
> since Pd at the time (and I think still) only does 32-bit computation in
> its "64-bit" version.

Today, Pd should work ok if you compile it for 64 bits (double) precision, but that is still experimental. And, anyway, I missed your reply by a handful of seconds, but as I suspected, I think there's a confusion wether we're talking about double precision in audio computations. We're not! We just want this library to run in the Pd version compiled for 64 bit processors, even though it still does only computations in single (32 bits) precision. I hope it is clear.

I have another issue where I wasn't even able to compile and run this library for Pd compiled for 32 bit processors, see https://github.com/ericlyon/fftease3.0-32bit/issues/14

But I think it is more important these days to make it work and compile it for 64 bits (and as for 32 bits it should also work just fine with the same code). BTW, for mac, a single compiled binary will run for Pd compiled for both 32 and 64 bit processors.

**Comment by ericlyon** [comment]
Yes, I understand. IIRC I successfully compiled the externals for Pd for 64-bit processors. I think the issue was that Miller had introduced a new way of accessing arrays, and discontinued an older function call in Pd-64bit. Anyway, I'm pretty sure the distro I shared has the code updates. At least I think I had it working on Mac OS at one point.

In any case, the newer distro to check is here: https://github.com/ericlyon/lyonpotpourri3.0
Please note that it's probably experimental.

**Comment by Lucarda** [comment]
Cool, all good!

I have uploaded to Deken `lyonpotpourri3.0` about an year ago and it is working Ok.

```
lyonpotpourri[v0.3](Windows-amd64-32)(Sources).dek
	Uploaded by lucarda @ 2019-11-19 04:34:34
```

Here goes a Zip file if you want to include it on your Website:

[lyonpotpourri[v0.3]Windows64.zip](https://github.com/ericlyon/fftease3.0-32bit/files/5593345/lyonpotpourri.v0.3.Windows-amd64-32.Sources.zip)

:)

**Comment by Lucarda** [comment]
Reopening.

I was having a chat with @porres.

I seem to got it wrong and thought lyonpotpourri3.0 was a newer replacement for fftease.

So there are plans for having fftease for 64bit ?

**Comment by porres** [comment]
> I don't have any problem with PRs, but I also basically have no idea how
> Github works, so I might need some advice about how to incorporate any code
> suggestions.

it's pretty simple and straight forward, if I can do it, anyone can :) 

when we fork the repository and make changes, we can send code revisions as Pull Requests here on github and they're shown in https://github.com/ericlyon/fftease3.0-32bit/pulls

By the way, there's onde there from @umlaeute and you could check it and I think it's a no brainer to accept and merge, see https://github.com/ericlyon/fftease3.0-32bit/pull/2

when you're there and logged in here on github, all you need to do is look for and click on the "merge" button. Then the changes get incorporated.

**Comment by ericlyon** [comment]
Just looked at that. I'd like to test out umlaeute's fork, but I couldn't
merge directly. There's some conflict that sends me to the command line:

git checkout -b umlaeute-more_nameclash_fixes mastergit pull
https://github.com/umlaeute/pd-fftease.git more_nameclash_fixes


After executing the second command, I got this error:
===
eric>git pull https://github.com/umlaeute/pd-fftease.git
more_nameclash_fixes

From https://github.com/umlaeute/pd-fftease
 * branch            more_nameclash_fixes -> FETCH_HEAD
error: Your local changes to the following files would be overwritten by
merge:
pvoc~.c
Please commit your changes or stash them before you merge.
Aborting
===
Any idea how to proceed?

Thanks,

Eric

**Comment by umlaeute** [comment]
i've updated my PR so it can now be cleanly merged into *current*¹ `master`.

¹ *current* as of 25th of November 2020; i don't necessarily expect it to be mergeable cleanly in 5 years from now.

**Comment by umlaeute** [comment]
btw (and just to make clear there is no confusion) my PR #2 (which just got included; wow @ericlyon that was very fast!) is unrelated to make fftease work on 64bit systems. (and there's still work to be done).

and while lyonpotpourri is in a better state (regarding 64bit pointers), it is also not fully ready yet.

it is *not* enough to just re-compile any of the two libs. code changes are required!

>  I think the issue was that Miller had introduced a new way of accessing arrays, and discontinued an older function call in Pd-64bit. 

ah yes. that "new way" was introduced in 2006. the old function call was deprecated at the same time. it was not explicitely "discontinued", it just happens to not work on 64bit systems, and that's unfixable (without changing the way Pd aligns its data arrays; which would have broken all binaries, regardless of whether they are for 32bit OSs or 64bit OSs or single- or double-precision).

**Comment by umlaeute** [comment]
@Lucarda wrote
> I'm trying to compile this for Windows 64bit and I can but I get warnings in the Pd console like:
>
>> "an operation on the array 'enrich-buffer' in the patch 'enrich~-help.pd' failed since it uses garray_getfloatarray while running 64-bit"

> and when closing help patches while DSP is on I get crashes.
> 
> Is this nice lib intended to not support 64bit Pd's?

this is actually a duplicate of #8

**Comment by ericlyon** [comment]
Since there seems to be some interest in other people contributing to update FFTease for 64-bit Pd builds, first of all, that's great, and I'll support the effort. I'll also state my constraints, so that there are no misunderstandings. Max/MSP is my primary development platform in the Max/Pd world. The Pd versions of LyonPotpourri and FFTease are forks, offered to the Pd community in the hope that they will be useful. Those packages have gone through a few different iterations, most recently with a split from a hybrid Max/Pd codebase that I developed around 2003. As Max 6 matured, I decided to split the code bases into a Max version and a Pd version. The Max API had simply drifted too far from Pd, and it no longer made sense to maintain a hybrid version. 

Sometime later, I had a reason to look at the 64-bit build of Pd, and it became clear that some code modifications would be required to compile my earlier Pd code for the Pd 64-bit build. At that time, I did update LyonPotpourri, but did not have time to update FFTease. In part, I was waiting for the 64-bit build of Pd to adopt double-precision, which has still not happened. Updating the Max versions to double-precision (Max 6) was a major hassle, so I didn't want to do two different rounds of updates for Pd. 

In any case, those considerations are now moot. I just don't have time for serious code development on Pd anytime soon. I did put the code out on Github, in case anyone was interested, but still don't know even the most rudimentary aspects of working with Github in a collaborative manner. I'm happy if I can get my own commits registered. But if other people want to bring the code up to date for 64-bit Pd, and add other fixes, again, I'm happy to support that effort. I just might not know how to do so technically on Github, which is why I asked for advice about merging. I'l also have to figure out how to test proposed updates before merging them. I'm sure it's extremely simple, but I don't have time to look into it, due to multiple other high-priority (for me) projects.

So again, thanks to anyone who would like to contribute to updating FFTease for Pd. Let's see if this goes anywhere.

**Comment by porres** [comment]
cool, I guess we can send you PRs and hopefully there won't be any hassle in merging them, just clicking the button on the website. I can help testing, so can @Lucarda, and @umlaeute is the wizard who's also willing to send fixes. BTW, iohannes, you mentioned you would eventually send a fix to https://github.com/ericlyon/fftease3.0-32bit/issues/8, is there a chance you could do that soon? If not, I could give it a try.

**Comment by Lucarda** [comment]
@ericlyon  I'd built for Windows64-bit and the issues seems gone ('garray_getfloatarray while running 64-bit' and the 'crashes when closing help patches).

:)

I'd make a small PR #23 that moves xy window position of the help patches to a more user friendly area on standard screens.

Wow, it got merged as I write.

:)

**Comment by ericlyon** [comment]
It's great that all of these improvements are happing so quickly. Thanks for the updates!

That said, the help files are still pretty terrible. They're more for testing the externals than actually explaining how to use them. That's something I may eventually work on.

**Comment by porres** [comment]
I usually like working with help files :) I could help... maybe after all the new improvements, we could make it 3.0.1 or something?

**Comment by ericlyon** [comment]
> I usually like working with help files :) I could help... maybe after all the new improvements, we could make it 3.0.1 or something?

Thanks for the offer! I may make one example help file, just to suggest a format, and then it would be great if you'd like to run with that model. Stay tuned...

**Comment by ericlyon** [comment]
@porres I'm attaching a sample help file here for drown~. The main criteria I'd suggest for good help files are:

1. FFTease icon at top left
2. Brief text description of object below icon
3. Every user-accessible method is shown and explained in the main patch
4. Logical breakdown into subpatches, for the more complicated objects

There are fairly good existing help files for the Max/MSP version of FFTease. They could serve as a model. 
( http://disis.music.vt.edu/eric/main/fftease/ )

Thanks!

[drown~-help.pd.zip](https://github.com/ericlyon/pd-fftease/files/5605671/drown.-help.pd.zip)

**Comment by porres** [comment]
cool, I can get to it by xmas time ;)

**Comment by porres** [comment]
Well, this thread is now a bit too vague, about the last thing on the documentation, I created a new issue, see https://github.com/ericlyon/pd-fftease/issues/26

Now, what about the rest, where do we stand? What's the next step? I guess we can open new issues for each thingy...

**Comment by porres** [comment]
@Lucarda , as you opened the issue, are you able to build just fine for windows?

**Comment by Lucarda** [comment]
Yes. All is good exept i get a dup of issue #7

**Comment by porres** [comment]
> Yes. All is good exept i get a dup of issue #7

so we talk about it over there :)

**Comment by porres** [comment]
@Lucarda, can we close this issue then?

And if so, I think we can have some builds and upload them to deken. I could open a new issue about uploading but I'll keep it here for now. One thing is that it seems there are still fixes to be made, such as #7 mentioned here, so maybe we want to settle that before we can say the build is over and make the release to deken?

I guess we'll always find bugs and other issues. I'm now also working on the documentation and it seems I'll find few things on the way. Nonetheless, I was hoping we could tag an official 3.0 release and upload it to deken. For other fixes and a rewrite of the documentation, we can just upload later under a new release version above 3.0, with a detailed changelog.

Now, I'm assuming the codebase has diverged from the original  MAX/MSP development, so both projects will have different bugs, different fixes and different versions and releases. Is this true and ok @ericlyon?

Note that FFTease (and lyonpotpourri) is currently being used in two forks of Pd. Pd-ceammc and Purr Data, so it's good we make an official 3.0 release as it is, which is already different than what they have, and keep on going from here, in a way they also know when to upgrade to a new release (3.0.1 / 3.1, whatever).

My main concern and interest here right now is that we have an upload to deken so FFTease (and later lyonporpourri) is also available for Pd Vanilla.

cheers

**Comment by ericlyon** [comment]
> @Lucarda, can we close this issue then?
> 
> And if so, I think we can have some builds and upload them to deken. I could open a new issue about uploading but I'll keep it here for now. One thing is that it seems there are still fixes to be made, such as #7 mentioned here, so maybe we want to settle that before we can say the build is over and make the release to deken?
> 
> I guess we'll always find bugs and other issues. I'm now also working on the documentation and it seems I'll find few things on the way. Nonetheless, I was hoping we could tag an official 3.0 release and upload it to deken. For other fixes and a rewrite of the documentation, we can just upload later under a new release version above 3.0, with a detailed changelog.
> 
> Now, I'm assuming the codebase has diverged from the original MAX/MSP development, so both projects will have different bugs, different fixes and different versions and releases. Is this true and ok @ericlyon?

That is correct.

> 
> Note that FFTease (and lyonpotpourri) is currently being used in two forks of Pd. Pd-ceammc and Purr Data, so it's good we make an official 3.0 release as it is, which is already different than what they have, and keep on going from here, in a way they also know when to upgrade to a new release (3.0.1 / 3.1, whatever).
> 
> My main concern and interest here right now is that we have an upload to deken so FFTease (and later lyonporpourri) is also available for Pd Vanilla.
> 

Good idea. I'll need to update the welcome text, before that's ready to go. Also, would want to address any issues about documenting installation procedures.

Thanks!
Eric

**Comment by Lucarda** [comment]
> @Lucarda, can we close this issue then?

> And if so, I think we can have some builds and upload them to deken. I could open a new issue about uploading but I'll keep it here for now. One thing is that it seems there are still fixes to be made, such as #7 mentioned here, so maybe we want to settle that before we can say the build is over and make the release to deken?


This issue can be closed. #7 is the main "Show stopper" for a Deken release.

**Comment by Lucarda** [comment]
Just to inform here that I compiled (as of 5cab997) on Windows64 and #7 is fixed (no crashes on mindwarp~ help). Also going through all help patches I get no errors/warnings in the console. 

:)

**Comment by porres** [comment]
cool, I already have a macOS build which is basically ready for a deken release => https://github.com/porres/pd-fftease/releases/tag/v3.0-0

I guess we only need to maybe change and update the readme and change/deal with the license because of https://github.com/ericlyon/pd-fftease/issues/3 but the binaries builds are fine

**Comment by ericlyon** [comment]
There are still some externals missing the "fftinfo" message. It would be good to get them updated so that the package is consistent. We will require an update on the readme, which I'll take care of. Also specific installation guidelines. Also, we need to remove the GPL dependency before publication. Also, need to get a full set of new help files. But lots of good progress so far. Thanks!

**Comment by porres** [comment]
As the one onboard with the new help files, it'll just take a while as I wanna really turn the all inside out, and i'm sure many bugs and issues will pop out then. A couple aklready emerged and we took care of that and I was also aware fftinfo was missing in other externals, for instance. 

I just think this is good enough and we could upload this as it so it is not too divergent from what's already being distributed in Pd-ceammc and Purr Data like I said above. If we make an official "3.0" release as it is, it's already different than what they have in those forks of Pd, but not too much... we can keep on going ffurther in a way they also know when to upgrade to a new release (3.0.1 / 3.1, whatever). But it'd be good to tell them to pair up to this "official" 3.0

anyway, that's just my 2 cents...

**Comment by ericlyon** [comment]
I'm fine with moving forward before you complete the deluxe new version of the help files. That being said, some of the help files are a mess, and I'd like to get them cleaned up before a new release.

**Comment by porres** [comment]
> some of the help files are a mess, and I'd like to get them cleaned up before a new release.

agreed, I thought about making a not so deluxe version, but I have a hard time doing such things and not going to the full extent. But I could give it a shot and work on something that's better quickly.

**Comment by porres** [comment]
> something that's better quickly.

and not far from max's help

**Comment by ericlyon** [comment]
That would be awesome, thanks!
